### PR TITLE
Migrate Intel 2021.3 build to gen-config

### PR DESCRIPTION
**Pull request by jmlapre**
@trilinos/framework 

## Motivation

We're moving to C++17.

## Related Issues

https://sems-atlassian-son.sandia.gov/jira/browse/TRILFRAME-93

## Testing

Manually ran GenConfig, cmake, and make.  Iterated upon errors.

Only two of the tests failed:
* STKUnit_tests_stk_mesh_unit_tests_MPI_4 (Failed)
* PanzerAdaptersSTK_tQuad8ToQuad4Factory_MPI_2 (Failed)

### CODE REVIEW

**Comment by rppawlo** [comment]
@jmlapre - can you post the output from the panzer failure (or email)? The failing test is on the critical path for an application code. Hopefully it is a simple fix.

**Comment by jmlapre** [comment]
> @jmlapre - can you post the output from the panzer failure (or email)? The failing test is on the critical path for an application code. Hopefully it is a simple fix.

```
(rhel7_sems-intel-2021.3-sems-openmpi-4.0.5) $ ctest -R PanzerAdaptersSTK_tQuad8ToQuad4Factory
Test project /scratch/jmlapre/Trilinos/build
    Start 3160: PanzerAdaptersSTK_tQuad8ToQuad4Factory_MPI_2
1/1 Test #3160: PanzerAdaptersSTK_tQuad8ToQuad4Factory_MPI_2 ...***Failed    2.21 sec

0% tests passed, 1 tests failed out of 1

Subproject Time Summary:
Panzer    =   4.41 sec*proc (1 test)

Total Test time (real) =   5.36 sec

The following tests FAILED:
	3160 - PanzerAdaptersSTK_tQuad8ToQuad4Factory_MPI_2 (Failed)
Errors while running CTest
Output from these tests are in: /scratch/jmlapre/Trilinos/build/Testing/Temporary/LastTest.log
Use "--rerun-failed --output-on-failure" to re-run the failed cases verbosely.
(rhel7_sems-intel-2021.3-sems-openmpi-4.0.5) $ ctest --rerun-failed --output-on-failure -R PanzerAdaptersSTK_tQuad8ToQuad4Factory
Test project /scratch/jmlapre/Trilinos/build
    Start 3160: PanzerAdaptersSTK_tQuad8ToQuad4Factory_MPI_2
1/1 Test #3160: PanzerAdaptersSTK_tQuad8ToQuad4Factory_MPI_2 ...***Failed    2.18 sec
Teuchos::GlobalMPISession::GlobalMPISession(): started processor with name ascicgpu050 and rank 1!
Teuchos::GlobalMPISession::GlobalMPISession(): started processor with name ascicgpu050 and rank 0!

***
*** Unit test suite ...
***


Sorting tests by group name then by the order they were added ... (time = 1.82e-06)

Running unit tests ...

IOSS: Adding property 'DECOMPOSITION_METHOD' with value 'rib'
0. tQuad8ToQuad4Factory_periodic_input_UnitTest ...
Exodus Library Warning/Error: [ex_open_par_int]
	EXODUS: ERROR: Attempting to open the NetCDF file:
	'quad8.gen'
	The NetCDF library was not built with PNetCDF support as required for parallel access to this file.

	NetCDF: Unknown file format

Exodus Library Warning/Error: [ex_open_par_int]
	ERROR: failed to open quad8.gen for read/write. Either the file does not exist,
	or there is a permission or file format issue.
	NetCDF: Unknown file format

Exodus Library Warning/Error: [ex_open_par_int]
	EXODUS: ERROR: Attempting to open the NetCDF file:
	'quad8.gen'
	The NetCDF library was not built with PNetCDF support as required for parallel access to this file.

	NetCDF: Unknown file format

Exodus Library Warning/Error: [ex_open_par_int]
	ERROR: failed to open quad8.gen for read/write. Either the file does not exist,
	or there is a permission or file format issue.
	NetCDF: Unknown file format
ERROR: Unable to open input exodus database file 'meshes/quad8.gen' on processors:
	0, 1


 p=0: *** Caught standard std::exception of type 'std::runtime_error' :

  Expr '!(Teuchos::is_null(m_database) || !m_database->ok(true))' eval'd to true, throwing.
  Error occurred at: stk_io/stk_io/InputFile.cpp:144
   0# 0x0000000001426916 in /scratch/jmlapre/Trilinos/build/packages/panzer/adapters-stk/test/stk_interface_test/PanzerAdaptersSTK_tQuad8ToQuad4Factory.exe
   1# 0x0000000000AEC9DC in /scratch/jmlapre/Trilinos/build/packages/panzer/adapters-stk/test/stk_interface_test/PanzerAdaptersSTK_tQuad8ToQuad4Factory.exe
   2# 0x0000000000AC1C89 in /scratch/jmlapre/Trilinos/build/packages/panzer/adapters-stk/test/stk_interface_test/PanzerAdaptersSTK_tQuad8ToQuad4Factory.exe
   3# 0x000000000046B7A5 in /scratch/jmlapre/Trilinos/build/packages/panzer/adapters-stk/test/stk_interface_test/PanzerAdaptersSTK_tQuad8ToQuad4Factory.exe
   4# 0x000000000046A685 in /scratch/jmlapre/Trilinos/build/packages/panzer/adapters-stk/test/stk_interface_test/PanzerAdaptersSTK_tQuad8ToQuad4Factory.exe
   5# 0x0000000000459661 in /scratch/jmlapre/Trilinos/build/packages/panzer/adapters-stk/test/stk_interface_test/PanzerAdaptersSTK_tQuad8ToQuad4Factory.exe
   6# 0x000000000209A88C in /scratch/jmlapre/Trilinos/build/packages/panzer/adapters-stk/test/stk_interface_test/PanzerAdaptersSTK_tQuad8ToQuad4Factory.exe
   7# 0x000000000209FF69 in /scratch/jmlapre/Trilinos/build/packages/panzer/adapters-stk/test/stk_interface_test/PanzerAdaptersSTK_tQuad8ToQuad4Factory.exe
   8# 0x000000000209CD3B in /scratch/jmlapre/Trilinos/build/packages/panzer/adapters-stk/test/stk_interface_test/PanzerAdaptersSTK_tQuad8ToQuad4Factory.exe
   9# 0x000000000209BB71 in /scratch/jmlapre/Trilinos/build/packages/panzer/adapters-stk/test/stk_interface_test/PanzerAdaptersSTK_tQuad8ToQuad4Factory.exe
  10# 0x0000000000461C6E in /scratch/jmlapre/Trilinos/build/packages/panzer/adapters-stk/test/stk_interface_test/PanzerAdaptersSTK_tQuad8ToQuad4Factory.exe
  11# __libc_start_main in /lib64/libc.so.6
  12# 0x0000000000459529 in /scratch/jmlapre/Trilinos/build/packages/panzer/adapters-stk/test/stk_interface_test/PanzerAdaptersSTK_tQuad8ToQuad4Factory.exe

  Error: ERROR: Could not open database 'meshes/quad8.gen' of type 'exodusii'

 NOTE: Unit test failed on all processes!
 [FAILED]  (0.00315 sec) tQuad8ToQuad4Factory_periodic_input_UnitTest
 Location: /scratch/jmlapre/Trilinos/packages/panzer/adapters-stk/test/stk_interface_test/tQuad8ToQuad4Factory.cpp:67


The following tests FAILED:
    0. tQuad8ToQuad4Factory_periodic_input_UnitTest ...

Total Time: 0.00327 sec

Summary: total = 1, run = 1, passed = 0, failed = 1

End Result: TEST FAILED
--------------------------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
--------------------------------------------------------------------------
--------------------------------------------------------------------------
mpiexec detected that one or more processes exited with non-zero status, thus causing
the job to be terminated. The first process to do so was:

  Process name: [[41092,1],1]
  Exit code:    1
--------------------------------------------------------------------------


0% tests passed, 1 tests failed out of 1

Subproject Time Summary:
Panzer    =   4.37 sec*proc (1 test)

Total Test time (real) =   5.38 sec

The following tests FAILED:
	3160 - PanzerAdaptersSTK_tQuad8ToQuad4Factory_MPI_2 (Failed)
Errors while running CTest
```

**Comment by rppawlo** [comment]
@jmlapre - thanks for uploading the output. It looks like this build is not loading the right TPL module for netcdf. We need the netcdf module that is built with pnetcdf enabled. That should clean up this error.

**Comment by jmlapre** [comment]
> @jmlapre - thanks for uploading the output. It looks like this build is not loading the right TPL module for netcdf. We need the netcdf module that is built with pnetcdf enabled. That should clean up this error.

It seems to be linked to both libraries already:

```
(rhel7_sems-intel-2021.3-sems-openmpi-4.0.5) $ ldd ./packages/panzer/adapters-stk/test/stk_interface_test/PanzerAdaptersSTK_tQuad8ToQuad4Factory.exe | grep -i net
	libpnetcdf.so.4 => /projects/sems/install/rhel7-x86_64/sems-dev/tpl/parallel-netcdf/1.12.2/intel/2021.3/openmpi/4.0.5/zcxigs7/lib/libpnetcdf.so.4 (0x00007fd1c7c28000)
	libnetcdf.so.19 => /projects/sems/install/rhel7-x86_64/sems-dev/tpl/netcdf-c/4.8.1/intel/2021.3/openmpi/4.0.5/3xp45tb/lib/libnetcdf.so.19 (0x00007fd1c78af000)
```

Any recommendations??

**Comment by e10harvey** [comment]
> > @jmlapre - thanks for uploading the output. It looks like this build is not loading the right TPL module for netcdf. We need the netcdf module that is built with pnetcdf enabled. That should clean up this error.
> 
> It seems to be linked to both libraries already:

This libnetcdf.so was **not** built with parallel support but CMake detects that `TPL_NetCdf_Parallel` is True. @jmlapre can you elaborate? @bartlettroscoe, @rppawlo: Do we need to tell the Trilinos CMake to use `PNetCdf` by setting `TPL_NetCdf_Parallel=OFF` in config-specs.ini?

**Comment by rppawlo** [comment]
I think you need to load the right module for netcdf - the one that was built with pnetcdf enabled. For example, in the sems-archive modules, you needed 
```
module load sems-archive-netcdf/4.7.3/parallel
```
not this one
```
sems-archive-netcdf/4.7.3/base
```
I just pulled this example from the atdm build scripts using teh archived sems modules. I don't know what modules are being loaded in this build, but if you can't find one that works, maybe ask sems to rebuild the module with pnetcdf included?

**Comment by bartlettroscoe** [comment]
@jmlapre, a couple of suggestions for GitHub PRs (or Issues) like this:

1. Give the PR name a meaningful title like "Migrate Intel 2021.3 build to gen-config (TRILFRAME-93)".  The title "TRILFRAME-93" will not mean anything to 99% of Trilinos developers.
2. Put in a link back to the actual JIRA issue [TRILFRAME-93](https://sems-atlassian-son.sandia.gov/jira/browse/TRILFRAME-93) so one can navigate there.
3. In JIRA issue [TRILFRAME-93](https://sems-atlassian-son.sandia.gov/jira/browse/TRILFRAME-93), and a comment linking to this new GitHub PR.

**Comment by jmlapre** [comment]
> > > @jmlapre - thanks for uploading the output. It looks like this build is not loading the right TPL module for netcdf. We need the netcdf module that is built with pnetcdf enabled. That should clean up this error.
> > 
> > 
> > It seems to be linked to both libraries already:
> 
> This libnetcdf.so was **not** built with parallel support but CMake detects that `TPL_NetCdf_Parallel` is True. @jmlapre can you elaborate? @bartlettroscoe, @rppawlo: Do we need to tell the Trilinos CMake to use `PNetCdf` by setting `TPL_NetCdf_Parallel=OFF` in config-specs.ini?

`TPL_NetCdf_Parallel` was enabled because the `sems-dev-parallel-netcdf` module was loaded but libnetcdf.so (from the base netcdf module) was not built with parallel support.  I'm not entirely sure if this is even a supported configuration.  Rebuilding the module with parallel enabled should fix it.

File: packages/framework/ini-files/config-specs.ini [hunk 1]
```diff
@@ -2350,6 +2350,38 @@ opt-set-cmake-var Zoltan2_scotch_example_MPI_4_DISABLE BOOL : ON
...
+opt-set-cmake-var TPL_Netcdf_LIBRARIES STRING FORCE : -L${NETCDF_C_LIB|ENV}/lib;${NETCDF_C_LIB|ENV}/libnetcdf.so;${PARALLEL_NETCDF_LIB|ENV}/libpnetcdf.a
+opt-set-cmake-var TPL_HDF5_LIBRARIES STRING FORCE : ${HDF5_LIB|ENV}/libhdf5_hl.so;${HDF5_LIB|ENV}/libhdf5.a;${ZLIB_LIB|ENV}/libz.a;-ldl
+opt-set-cmake-var MueLu_UnitTestsEpetra_MPI_1_DISABLE BOOL FORCE : ON
+opt-set-cmake-var MueLu_UnitTestsEpetra_MPI_4_DISABLE BOOL FORCE : ON
+
+opt-set-cmake-var Trilinos_ENABLE_STKBalance BOOL FORCE : OFF
+opt-set-cmake-var MueLu_UnitTestsTpetra_MPI_1_DISABLE BOOL FORCE : ON
+opt-set-cmake-var MueLu_UnitTestsTpetra_MPI_4_DISABLE BOOL FORCE : ON
+
+# /JML
```
Comment (e10harvey) [inline packages/framework/ini-files/config-specs.ini:2383]:
```suggestion
```

**Review by e10harvey — CHANGES_REQUESTED** [review]
Thanks, @jmlapre ! Please see comments below.

File: packages/framework/ini-files/config-specs.ini [hunk 2]
```diff
@@ -2350,6 +2350,33 @@ opt-set-cmake-var Zoltan2_scotch_example_MPI_4_DISABLE BOOL : ON
...
+use USE-UVM|NO
+use USE-DEPRECATED|YES
+use PACKAGE-ENABLES|NO-PACKAGE-ENABLES
+use COMMON
+
+opt-set-cmake-var CMAKE_CXX_STANDARD STRING : 17
+opt-set-cmake-var ML_ENABLE_SuperLU BOOL FORCE : OFF
+opt-set-cmake-var TPL_Netcdf_LIBRARIES STRING FORCE : -L${NETCDF_C_LIB|ENV}/lib;${NETCDF_C_LIB|ENV}/libnetcdf.so;${PARALLEL_NETCDF_LIB|ENV}/libpnetcdf.a
+opt-set-cmake-var TPL_HDF5_LIBRARIES STRING FORCE : ${HDF5_LIB|ENV}/libhdf5_hl.so;${HDF5_LIB|ENV}/libhdf5.a;${ZLIB_LIB|ENV}/libz.a;-ldl
+opt-set-cmake-var MueLu_UnitTestsEpetra_MPI_1_DISABLE BOOL FORCE : ON
```
Comment (e10harvey) [inline packages/framework/ini-files/config-specs.ini:2373]: @jmlapre: Please link to the related github issue that reports these test failures to MueLu.

File: packages/framework/ini-files/config-specs.ini [hunk 3]
```diff
@@ -2350,6 +2350,33 @@ opt-set-cmake-var Zoltan2_scotch_example_MPI_4_DISABLE BOOL : ON
...
+opt-set-cmake-var MueLu_UnitTestsEpetra_MPI_4_DISABLE BOOL FORCE : ON
+
+opt-set-cmake-var Trilinos_ENABLE_STKBalance BOOL FORCE : OFF
```
Comment (e10harvey) [inline packages/framework/ini-files/config-specs.ini:2376]: Quick comment please.

File: packages/framework/ini-files/config-specs.ini (hunk 2 above)
Comment (jmlapre) [reply packages/framework/ini-files/config-specs.ini:2373]: I'm working through it.  Many of the disabled tests can be re-enabled as the related testcases are now passing.  Down to the last few now...

**Review by e10harvey — APPROVED** [review]
Thanks, @jmlapre !

Please ensure these changes get included in https://github.com/e10harvey/Trilinos/tree/sprint17_cpp17 and create a nightly run of `rhel7_sems-intel-2021.3-sems-openmpi-4.0.5_release-debug_shared_no-kokkos-arch_no-asan_no-complex_fpic_mpi_no-pt_no-rdc_no-uvm_deprecated-on_all`

**Comment by e10harvey** [comment]
@jmlapre: Please merge the commit that I approved previously into develop and then get these changes incorporated into the sprint17_cpp17.

**Comment by e10harvey** [comment]
> @jmlapre: Please merge the commit that I approved previously into develop and then get these changes incorporated into the sprint17_cpp17.

Alternatively, feel free to close this PR and get these change incorporated into sprint17_cpp17.

**Comment by e10harvey** [comment]
THanks, @jmlapre ! Closing in favor of https://github.com/e10harvey/Trilinos/pull/17.

**Pull request closed without merging**